### PR TITLE
fix(formations): stop drawer silently dropping assignee and due date

### LINKED ISSUE #2694: Formation item drawer: Save silently drops assignee and due date
## Summary

Saving the checklist item drawer with an assignee and due date entered leaves **neither field persisted** — with **no error anywhere**. The note in the same Save lands, gets the only history entry, and the drawer reports "Saved". Reproduced on production 17 Sep on "Formation review and packet" (`formation_review_packet`, project `stctf`).

> **Correction (same day):** the first version of this issue assumed the assignment request was sent and failed with an error toast. Production RUM disproved that — see the evidence below. The reporter saw no toast because **no assignment request was ever made**.

## Evidence, straight from prod RUM (17 Sep, usr "Nirav Patel")

```
19:18:17Z  POST  201  /api/projects/fc9130cf-…/permissions          ← reporter adds a permissions entry
19:18:40Z  PATCH 200  /api/formations/fc9130cf-…/items/formation_review_packet   ← the note leg. 200.
19:18:40Z  GET   200  /api/projects/stctf/formation                 ← section refresh
19:18:40Z  GET   200  …/items/formation_review_packet               ← drawer reload
20:23–20:29Z  GET 200 ×5  …/items/formation_review_packet, others   ← reopenings; ZERO writes
```

**There is no `POST …/assignment` anywhere in the session.** The Save diffed the form, saw no committed assignee/due-date change, sent the note alone, and said "Saved".

## Three defects, one experience

**1. The observed repro: entered values never commit, and everything about that is silent.**
An assignee typed into the picker only commits via a dropdown selection (`formation-item-drawer` + `lfx-user-search`). Worse, `lfx-user-search.onSearchBlur` **snaps the box back to the committed value on blur — and blur fires before the Save button's own click** — so the typed name visibly vanishes in the very gesture that saves, with no feedback. If the assignee/due date were the only edits, `onSaveDetails` hit its silent `if (nothing changed) return` — no request, no toast, nothing. The enabling condition is #2594: the picker searches `committee_member`, so a staff member (or any project grant holder without a committee seat) may have **no selectable result at all** — typed text is then the only thing they *can* enter, and it is always discarded.

**2. This surface never gates writer-only controls.** The drawer's `canWrite` input defaults `true` and `formation-checklist-section` never bound it (`FormationChecklistResponse` carried no per-caller writer flag to bind). Per the service owner's rendering contract, `assign`/`set_due_date` carry `requires_relation: "writer"` and the UI must filter on it; this surface doesn't consume `available_actions` at all. A non-writer is offered editable fields whose save can only 403. (The Me-lens drawer host binds a real `can_write`; this section — project page **and** queue drill-down — did not.)

**3. When the assignment write IS sent, a refused assignee silently costs the due date too.** Upstream accepts an assignee only from the project's direct `settings.Writers`/`settings.Auditors` (400 `assignee_not_on_project` otherwise — staff access via `lf-staff` ROOT inheritance doesn't count until #1954), and `item_assignment.go` returns **before** applying `DueDate`, so a combined body loses the due date with the refused assignee. The old UI sent exactly such a combined body, and mapped every failure to one generic toast (no 403 branch, no `assignee_not_on_project` copy).

## Fix (PR to follow)

- `lfx-user-search` records the text a blur snap-back discards; the drawer checks that record when Save is clicked, warns naming the exact text, and blocks that Save once (a deliberate second Save proceeds with what actually committed). Checked at save time rather than toasted at blur time because blur equally precedes a dropdown pick's own click — a blur-time notice would false-fire on every successful mouse selection.
- A Save with nothing committed now says "No changes to save" instead of doing nothing silently.
- The BFF checklist response now carries a per-caller `can_write` (single-project access check, fail-closed), and the section binds it into **both** the drawer (read-only fields + explainer) and the row (status menu, overflow menu, and gated quick action suppressed) — no writer-gated affordance is offered to a caller it can only 403 for.
- The Save now writes note → due date → assignee as **separate version-chained requests** (assignee last, as the one leg with a business-rule refusal), keeps whatever landed, names the saved, failed, **and unattempted** legs — with specific copy for `ASSIGNEE_NOT_ON_PROJECT` and 403.

Left open here: the picker corpus itself (#2594 — until a project-grant-holder corpus exists, a valid pick may simply be impossible), and the service-side question of whether `assign_item` should apply `due_date` independently of a refused assignee (owner's call; the UI no longer depends on it).

Also observed in the same sessions, out of scope here: `PATCH …/items/*/status` and skip writes returning 400 surfaced only as generic toasts + console errors (13:51Z, 16:37Z, 18:42Z, 20:29Z). May deserve its own look.

Rendering contract reference: https://gist.github.com/prabodhcs/d1ce8940003f2e09c91570e15dc56bee

Refs #1958, #2594, #1954. Part of epic #1965.

### PR DESCRIPTION
**Pull request by niravpatel27**
## Summary

Saving the formation item drawer with an assignee and due date entered could leave **neither persisted, with no feedback anywhere** — the note leg landed, the drawer said "Saved", and the fields read empty on reopen. Verified against production RUM (17 Sep): the Save sent **only the note PATCH**; no assignment request was ever made. Three stacked defects, all fixed here:

1. **Entered values never committed, silently.** An assignee typed into the picker only commits via a dropdown pick, and `lfx-user-search` snaps the box back on blur — which fires before Save's own click — so the typed name vanished in the very gesture that saved. A Save whose diff then saw nothing returned without a request or a toast. The search now **records** the discarded text (recorded, not toasted at blur: blur equally precedes a dropdown pick's own click, so a blur-time notice would false-fire on every successful mouse selection); the drawer consumes the record on Save, warns naming the exact text, and blocks that Save once — a deliberate second Save proceeds with what actually committed. An all-no-op Save now says "No changes to save".
2. **Writer-gated controls were offered to every caller.** The drawer's `canWrite` input defaulted `true` and the checklist section never bound it. `FormationChecklistResponse` now carries a per-caller `can_write` — resolved via the single-project access check (never batch, LFXV2-2823), fail-closed — and the section binds it into **both** the drawer (read-only assignee/due-date fields with the existing explainer) and the row (status menu, overflow menu, and gated quick action suppressed, mirroring `readOnly`). Navigation and the auditor-editable notes stay.
3. **A refused assignee silently discarded the due date sent beside it.** Upstream rejects a combined assignment body wholesale on `assignee_not_on_project`. Save now writes note → due date → assignee as **separate version-chained requests** (assignee last, as the one leg with a business-rule refusal), keeps whatever landed, and names the saved, failed, and unattempted legs — with specific copy for `ASSIGNEE_NOT_ON_PROJECT` and 403 instead of the old single generic toast.

## Changes

- `packages/shared`: `FormationChecklistResponse.can_write` (per-caller, fail-closed; documented).
- BFF `formation.service.ts`: `getProjectFormation` swaps the access-less cached project read for the access-checked one and returns `can_write` — serves both checklist routes (project page + queue drill-down) via the shared controller.
- `formation-item-drawer`: split version-chained save legs; save-time discarded-assignee guard; no-op Save feedback; partial-failure toasts naming saved/failed/unattempted legs; reason-specific error copy.
- `formation-checklist-section` / `formation-checklist-row`: `can_write` bound into drawer and row; row mutation surfaces gated like `readOnly`.
- `lfx-user-search`: discarded-text record + `consumeDiscardedText()`; superseded by any later pick, clear, or fresh typing.
- E2E mocks: every hand-rolled checklist body now carries `can_write: true`.

## Testing

- Full unit suites green: 3,522 app tests + server suite (`yarn test`), including new specs for every behavior above (split legs + version chaining, partial/unattempted reporting, discard record semantics incl. the mouse-pick supersede, row/drawer `can_write` gating, BFF `can_write` resolution fail-closed).
- `yarn lint` (0 errors), `yarn build` pass; local whole-branch review batch (general + repo conventions + learnings) run and all findings addressed in the second commit.

## External references

- Formation service rendering contract (service owner): https://gist.github.com/prabodhcs/d1ce8940003f2e09c91570e15dc56bee — `assign`/`set_due_date` carry `requires_relation: "writer"`; the UI must filter on standing it already holds.
- Left open deliberately: the picker corpus itself (#2594 — until a project-grant-holder corpus exists, a valid pick may simply be impossible; the new warn at least says so), staff-writer tuples (#1954), and the service-side question of whether `assign_item` should apply `due_date` independently of a refused assignee (owner's call; the UI no longer depends on it).

Closes #2694. Refs #2594, #1954.

### CODE REVIEW

**Comment by dealako** [comment]
Hey @niravpatel27, thanks for the thorough writeup — this is a well-scoped fix for a nasty class of silent-data-loss bug. The three-way root-cause breakdown (uncommitted picker text, unscoped `canWrite`, and the combined-request rollback) is clearly reasoned, and the version-chained save with per-leg success/failure/unattempted reporting directly addresses the original symptom. The `can_write` fail-closed resolution and the blur-vs-click race fix in `lfx-user-search` are both handled carefully, with test coverage for the tricky edge cases (version threading between legs, discard-record supersede on a later pick, fail-closed `can_write`).

Three independent reviewers (security/privacy, correctness/performance/tests, style/API/docs) covered this diff. No bot (CodeRabbit, Cursor) posted actionable inline findings — CodeRabbit skipped (auto-review disabled on this repo) and Cursor's comment was summary-only.

**Issue count:**
- 🔴 Blocking: 0
- 🟡 Minor: 0
- ⚪ Nit: 0
- ❔ Question: 1: sibling test for the "middle leg fails" case doesn't assert that reload fires afterward, unlike the adjacent "note succeeds, assignment fails" test which does assert it

**Final decision: ✅ Approved with minor comments**

File: apps/lfx-one/src/app/modules/dashboards/components/formation-item-drawer/formation-item-drawer.component.spec.ts [hunk 1]
```diff
@@ -559,6 +563,198 @@ describe('FormationItemDrawerComponent', () => {
...
+      // Second Save: the record was consumed by the warning — a deliberate repeat proceeds with
+      // what actually committed (the note), still without inventing an assignee.
+      (query('[data-testid="formation-item-drawer-save"] button') as HTMLElement)?.click();
+      await fixture.whenStable();
+
+      expect(updateFormationItemMock).toHaveBeenCalledWith(item.project_uid, item.template_item_key, String(item.version), { note: 'new note' });
+      expect(updateFormationItemAssignmentMock).not.toHaveBeenCalled();
+    });
+
+    it('names unattempted trailing legs when a middle leg fails — a failed due date must not silently drop the pending assignee', async () => {
```
Comment (dealako) [inline apps/lfx-one/src/app/modules/dashboards/components/formation-item-drawer/formation-item-drawer.component.spec.ts:682]: [question]

Issue: this test verifies the per-leg error message names the unattempted assignee leg, but doesn't assert that `reload$` actually fires after the partial failure — unlike the sibling test above ("reloads after the note leg succeeds but the assignment leg fails") which explicitly asserts `getFormationItemMock` is called twice.

Proof: the implementation's error handler (`formation-item-drawer.component.ts`) calls `reloadIfStillShowing(item)` unconditionally whenever `lastSavedItem` is truthy, so the reload should fire here too (the note leg landed before the due-date leg failed) — but this specific test never checks `getFormationItemMock`'s call count, so a regression in that branch wouldn't be caught here.

Why it matters: if a future change scoped the reload to only the assignment-leg failure path (rather than "any leg succeeded"), a retry after a failed due-date leg could resend against a stale version and 412 — and this test suite wouldn't catch it.

Fix: add `expect(getFormationItemMock).toHaveBeenCalledTimes(2)` here, mirroring the assertion in the test above.

**Review by dealako — APPROVED** [review]
Approving — solid fix for a real silent-data-loss bug, with careful handling of the version-chaining and blur/click race edge cases. One open question inline, non-blocking.

**Review by MRashad26 — CHANGES_REQUESTED** [review]
Ran a full review plus the `code-standards-enforcer` agent (`~/LFX/code-enforcer-agent.md`). 10 findings, most severity concentrated in the top two.

**Major — the PR's own fix (GH-2694) isn't applied to other call sites with the identical bug (`staff-edit-dialog.component.ts:121`).** This PR adds `discardedSearchText`/`consumeDiscardedText()` to `user-search.component.ts` to fix GH-2694 (typed-but-unselected assignee text silently discarded on blur before Save), and wires the fix into `formation-item-drawer`'s `onSaveDetails()`. But `staff-edit-dialog` binds the same `[displayValue]` pattern on the staff role picker and never calls `consumeDiscardedText()`. A writer can type a name into the role picker, not click a suggestion before clicking Save (the same blur-before-click race GH-2694 fixed elsewhere), and `onSubmit()` never consumes the discarded text — the email/name `FormControl`s stay at their old value, the PUT sends the unchanged assignee, and the UI reports "Role updated successfully" even though nothing changed.

**Major — same gap, second call site (`meeting-details.component.ts:282`).** The Meeting Organizer `lfx-user-search` also binds `[displayValue]` (so the same blur-discard race is live), but the save flow never calls `consumeDiscardedText()` either. A typed-but-unselected organizer name is silently discarded before save reads the form, `ownerEmail`/`ownerUsername` keep the previous organizer, and the save reports success with no indication the edit was dropped.

**Minor — read-only gating inconsistent with the component's own documented PrimeNG caveat (`formation-item-drawer.component.html:143`).** `assignmentReadOnly()` gates the assignee `lfx-user-search` via `[readonly]` only, but `user-search.component.html`'s own header comment documents that PrimeNG gates at least one affordance (the clear icon) on `[disabled]`, not `[readonly]` — a gap this diff doesn't audit for the dropdown-item-select path. A non-writer could visually land a new name in the box via the overlay even though the actual write is correctly suppressed by `canWrite()` in `onSaveDetails`, leaving confusing UI state until the drawer is reopened.

**Minor — new transient partially-applied state (`formations.route.ts:295`).** The due-date leg and assignee leg are now two sequential, separate `POST .../assignment` requests instead of one combined write. A concurrent reader refreshing between the two round trips could see the due date already updated but the old assignee still attached — a state impossible under the old single-request design. Self-heals on the next read, so severity is low, but it's new behavior worth being aware of.

**Minor — two independent encodings of the same upstream error contract (`formation-item-drawer.component.ts:536`).** The client hardcodes the literal `'ASSIGNEE_NOT_ON_PROJECT'` string while the server's `mapFormationWriteError` (`formation.service.ts:~972`) independently uppercases whatever raw reason string upstream returns, with no shared constant. If upstream renames or adds a variant of the reason string, the client's exact-string switch would silently stop matching and fall back to a generic error message, losing the more helpful guidance copy with no compile-time signal of the drift.

**Minor — duplicated claim-guard-free busy-state helpers (`formation-item-drawer.component.ts:643`).** `beginWrite`/`endWrite` duplicate `formation-checklist-section.component.ts`'s `beginSubmitting`/`endSubmitting` (~line 385) but drop the claim-guard (the section's version refuses to double-claim a uid; the drawer's doesn't). Not exploitable today since existing `busy()`/`mutationInFlight` gating already prevents concurrent writes on the same item, but a future guard-semantics fix applied to only one of the two hand-copied implementations would silently reintroduce the race.

**Minor — backend degraded-write signal is typed but never read (`formation-item-drawer.component.ts:415`).** The save chain is typed `Observable<FormationItemWriteResult>` (item + etag + item_state) and seeded with a synthetic `item_state: 'complete'`, but only `.item` is ever read anywhere in `onSaveDetails`. If a leg's post-write remap degrades (`item_state === 'stale'`), there's no code path to detect or surface it — the user is told the save succeeded with no warning the returned item may reflect a degraded/stale remap the backend already computed.

**Minor — fail-open write-access default with no shared source of truth (`formation-item-drawer.component.ts:68`).** `canWrite = input<boolean>(true)` is fail-open by default on both the row and drawer, and three independent "does this caller have write access" resolutions now exist codebase-wide (`formation-checklist-section`'s `can_write` computed, `dashboard-formation-item-drawer-host`'s `request.canWrite ?? true`, and `getMyFormationWork`'s `writerByProject` map). No live exploit today since all current hosts bind it explicitly, but any future host that forgets to bind `[canWrite]` would render a fully-writable component for a read-only caller with no visual signal anything is missing.

**Nit — redundant derivable state (`formation-item-drawer.component.ts:413`).** `failedIndex` is fully derivable from `savedLabels.length` (legs run strictly sequentially via chained `switchMap`, and a failure aborts the chain immediately, so `savedLabels.length` always equals the failed leg's index when `catchError` fires). No live bug, just one more variable a future refactor of the chain-building loop must keep in sync.

**Nit — nested ternary violates CLAUDE.md line 175 (`formation-item-drawer.component.ts:481`).** `unattempted.length > 0 ? \`...${unattempted.length > 1 ? 'were' : 'was'}...\` : ''` nests a ternary inside a ternary, violating this repo's explicit "Never nest ternary expressions" rule. The inner pluralization ternary should be extracted to a local const before the outer ternary.

Requesting changes primarily for the two Major findings — the PR's own stated fix (GH-2694) is incomplete, since the identical bug pattern is still live on two other call sites (`staff-edit-dialog`, `meeting-details`) that bind the same `[displayValue]` pattern.
